### PR TITLE
Improve resolving paths

### PR DESCRIPTION
**Pull request by zenangst**
This PR aims to improve the usability of resolving key paths.

With this pull request you can resolve nested paths like this;

```swift
XCTAssertEqual(json.path("school.clubs.0.detail.name"), "DC")
    XCTAssertEqual(json.path("school.clubs.0.detail.people.0.first_name"), "Clark")
    XCTAssertEqual(json.path("school.clubs.0.detail.people.0.age"), 78)
    XCTAssertEqual(json.path("school.clubs.0.detail.people")!, [
      [
        "first_name": "Clark",
        "last_name": "Kent",
        "age" : 78
      ],
      [
        "first_name": "Bruce",
        "last_name": "Wayne",
        "age" : 77
      ]
      ])
```

It can achieve this with the refactoring done in `PathAccessible` to introduce method overloading for the `path` method.

### CODE REVIEW

**Comment by zenangst** [comment]
This still needs some pushes to work properly.

File: Sources/Shared/Protocols/PathAccessible.swift [hunk 1]
```diff
@@ -36,19 +37,50 @@ public extension PathAccessible {
...
+
+  func path(path: [SubscriptKind]) -> JSONDictionary? {
+    return resolve(path)
+  }
+
+  func path(keyPath: String) -> JSONDictionary? {
+    return resolve(keyPath)
+  }
+
+  func path(keyPath: String) -> String? {
```
Comment (onmyway133) [inline Sources/Shared/Protocols/PathAccessible.swift]: We have `property, array, dictionary, ...`, what if we can have overloading functions like `access` that returns the corresponding types? It is like another convenient (and confused as well), and `resolve` can just call `access`

**Comment by zenangst** [comment]
@onmyway133 On the second thought, I think this should work 😁

File: Sources/Shared/Protocols/PathAccessible.swift [hunk 2]
```diff
@@ -36,19 +37,50 @@ public extension PathAccessible {
...
         return .Key($0)
       }
     }
 
-    return path(kinds)
+    return resolve(kinds)
+  }
+
+  private func extractKey(path: String) -> (key: String, keyPath: String) {
+    return (key: path.split(".").last!,
```
Comment (onmyway133) [inline Sources/Shared/Protocols/PathAccessible.swift]: Should we check instead of `!` ?

File: Sources/Shared/Protocols/PathAccessible.swift (hunk 1 above)
Comment (zenangst) [inline Sources/Shared/Protocols/PathAccessible.swift]: Yeah I thought about this as well, right now when working on a demo I use `property` and `path` but both could be done with one method that takes the paths into account and does the appropriate thing for the user instead of having multiple methods. The public API is way to big right now.

`array` and `dictionary` could become hidden methods used by `property`.

I think my vote is on `property` and not `path`. Properties can be resolved with keys or key paths. What do you think?

File: Sources/Shared/Protocols/PathAccessible.swift (hunk 2 above)
Comment (zenangst) [inline Sources/Shared/Protocols/PathAccessible.swift]: Good point, this can probably be improved :)

File: Sources/Shared/Protocols/PathAccessible.swift (hunk 1 above)
Comment (onmyway133) [inline Sources/Shared/Protocols/PathAccessible.swift]: @zenangst yeah, I think that should be less confusing

Comment (zenangst) [inline Sources/Shared/Protocols/PathAccessible.swift]: We could do that refactoring in a different pull request and then bump to a new major version number, right? 😁

**Comment by onmyway133** [comment]
@zenangst is it done for merging?

**Comment by zenangst** [comment]
@onmyway133 yeah I think so 😁 we can improve it other pull requests to not have this grow out of proportion.

**Comment by onmyway133** [comment]
🎸 🎵 🎵
